### PR TITLE
cephadm: verify config file exists when inferring it

### PR DESCRIPTION
**Pull request by adk3798**
Fixes: https://tracker.ceph.com/issues/54571

Signed-off-by: Adam King <adking@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -1873,7 +1873,12 @@ def infer_config(func: FuncT) -> FuncT:
             if not name:
                 daemon_list = list_daemons(ctx, detail=False)
                 for daemon in daemon_list:
-                    if daemon.get('name', '').startswith('mon.') and daemon.get('fsid', '') == ctx.fsid:
+                    if (
+                        daemon.get('name', '').startswith('mon.')
+                        and daemon.get('fsid', '') == ctx.fsid
+                        and daemon.get('style', '') == 'cephadm:v1'
+                        and os.path.exists(f'/var/lib/ceph/{ctx.fsid}/{daemon["name"]}/config')
```
[the remark this answers is not in the document]
Comment (mgfritch) [reply src/cephadm/cephadm:1880]: ... or maybe this ... ?

`data_dir = get_data_dir(fsid, ctx.data_dir, daemon_type, daemon_id)`

File: src/cephadm/tests/test_cephadm.py [hunk 2]
```diff
@@ -520,7 +527,8 @@ def test_infer_fsid(self, _call, fsid, ceph_conf, list_daemons, result, err, cep
             ),
         ])
     @mock.patch('cephadm.call')
-    def test_infer_config(self, _call, fsid, config, name, list_daemons, result, cephadm_fs):
+    @mock.patch('cephadm.logger')
+    def test_infer_config(self, logger, _call, fsid, config, name, list_daemons, result, cephadm_fs):
```
[the remark this answers is not in the document]
Comment (adk3798) [reply src/cephadm/tests/test_cephadm.py:531]: no, I'm just mocking it so calls to logger.info in the actual cephadm code being tested don't throw errors

File: src/cephadm/cephadm [hunk 3]
```diff
@@ -1873,11 +1873,16 @@ def infer_config(func: FuncT) -> FuncT:
             if not name:
                 daemon_list = list_daemons(ctx, detail=False)
                 for daemon in daemon_list:
-                    if daemon.get('name', '').startswith('mon.') and daemon.get('fsid', '') == ctx.fsid:
+                    if (
+                        daemon.get('name', '').startswith('mon.')
+                        and daemon.get('fsid', '') == ctx.fsid
+                        and daemon.get('style', '') == 'cephadm:v1'
+                        and os.path.exists(os.path.join(get_data_dir(ctx.fsid, ctx.data_dir, 'mon', daemon['name'].split('.', 1)[1]), 'config'))
```
Comment (phlogistonjohn) [inline src/cephadm/cephadm:1880]: This os.path.join(...) line and the one below on line 1885 seem rather complex to read to me. The similarity between the two and complexity suggest to me that a small function may be in order.  It's not a requirement, but I think future readers would be able to comprehend it better.

**Comment by adk3798** [comment]
http://pulpito.front.sepia.ceph.com/adking-2022-03-21_05:58:33-orch:cephadm-wip-adk-testing-2022-03-20-2156-distro-basic-smithi/

1 Dead job due to failing to power on a machine
2 Failed jobs due to unknown flag: --sort caused by another PR in the run
1 Failed job due to Error EINVAL: Invalid service 'ingress.nfs.test'. Use 'ceph orch ls' to list available services caused by another PR in the run

File: src/cephadm/cephadm (hunk 3 above)
Comment (adk3798) [reply src/cephadm/cephadm]: done

**Review by mgfritch — APPROVED** [review]

**Comment by adk3798** [comment]
http://pulpito.front.sepia.ceph.com/adking-2022-03-23_02:54:35-orch:cephadm-wip-adk-testing-2022-03-22-2000-distro-basic-smithi/

2 Failures caused by wrong error code from host add command due to another PR included in the run

**Pull request merged**